// bippy/dist/index.cjs
'use strict';

require('./chunk-62AO6VBI.cjs');
var chunkXUTTH4PR_cjs = require('./chunk-XUTTH4PR.cjs');
require('./chunk-AQ674A4M.cjs');



Object.defineProperty(exports, "BIPPY_INSTRUMENTATION_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.BIPPY_INSTRUMENTATION_STRING; }
});
Object.defineProperty(exports, "CONCURRENT_MODE_NUMBER", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.CONCURRENT_MODE_NUMBER; }
});
Object.defineProperty(exports, "CONCURRENT_MODE_SYMBOL_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.CONCURRENT_MODE_SYMBOL_STRING; }
});
Object.defineProperty(exports, "ClassComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.ClassComponentTag; }
});
Object.defineProperty(exports, "ContextConsumerTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.ContextConsumerTag; }
});
Object.defineProperty(exports, "DEPRECATED_ASYNC_MODE_SYMBOL_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.DEPRECATED_ASYNC_MODE_SYMBOL_STRING; }
});
Object.defineProperty(exports, "DehydratedSuspenseComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.DehydratedSuspenseComponentTag; }
});
Object.defineProperty(exports, "ELEMENT_TYPE_SYMBOL_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.ELEMENT_TYPE_SYMBOL_STRING; }
});
Object.defineProperty(exports, "ForwardRefTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.ForwardRefTag; }
});
Object.defineProperty(exports, "FragmentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.FragmentTag; }
});
Object.defineProperty(exports, "FunctionComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.FunctionComponentTag; }
});
Object.defineProperty(exports, "HostComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.HostComponentTag; }
});
Object.defineProperty(exports, "HostHoistableTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.HostHoistableTag; }
});
Object.defineProperty(exports, "HostRootTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.HostRootTag; }
});
Object.defineProperty(exports, "HostSingletonTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.HostSingletonTag; }
});
Object.defineProperty(exports, "HostTextTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.HostTextTag; }
});
Object.defineProperty(exports, "INSTALL_ERROR", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.INSTALL_ERROR; }
});
Object.defineProperty(exports, "INSTALL_HOOK_SCRIPT_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.INSTALL_HOOK_SCRIPT_STRING; }
});
Object.defineProperty(exports, "LegacyHiddenComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.LegacyHiddenComponentTag; }
});
Object.defineProperty(exports, "MemoComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.MemoComponentTag; }
});
Object.defineProperty(exports, "OffscreenComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.OffscreenComponentTag; }
});
Object.defineProperty(exports, "SimpleMemoComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.SimpleMemoComponentTag; }
});
Object.defineProperty(exports, "SuspenseComponentTag", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.SuspenseComponentTag; }
});
Object.defineProperty(exports, "TRANSITIONAL_ELEMENT_TYPE_SYMBOL_STRING", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.TRANSITIONAL_ELEMENT_TYPE_SYMBOL_STRING; }
});
Object.defineProperty(exports, "_fiberRoots", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs._fiberRoots; }
});
Object.defineProperty(exports, "createFiberVisitor", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.createFiberVisitor; }
});
Object.defineProperty(exports, "detectReactBuildType", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.detectReactBuildType; }
});
Object.defineProperty(exports, "didFiberCommit", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.didFiberCommit; }
});
Object.defineProperty(exports, "didFiberRender", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.didFiberRender; }
});
Object.defineProperty(exports, "fiberIdMap", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.fiberIdMap; }
});
Object.defineProperty(exports, "getDisplayName", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getDisplayName; }
});
Object.defineProperty(exports, "getFiberFromHostInstance", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getFiberFromHostInstance; }
});
Object.defineProperty(exports, "getFiberId", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getFiberId; }
});
Object.defineProperty(exports, "getFiberStack", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getFiberStack; }
});
Object.defineProperty(exports, "getLatestFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getLatestFiber; }
});
Object.defineProperty(exports, "getMutatedHostFibers", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getMutatedHostFibers; }
});
Object.defineProperty(exports, "getNearestHostFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getNearestHostFiber; }
});
Object.defineProperty(exports, "getNearestHostFibers", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getNearestHostFibers; }
});
Object.defineProperty(exports, "getRDTHook", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getRDTHook; }
});
Object.defineProperty(exports, "getTimings", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getTimings; }
});
Object.defineProperty(exports, "getType", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.getType; }
});
Object.defineProperty(exports, "hasMemoCache", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.hasMemoCache; }
});
Object.defineProperty(exports, "hasRDTHook", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.hasRDTHook; }
});
Object.defineProperty(exports, "installRDTHook", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.installRDTHook; }
});
Object.defineProperty(exports, "instrument", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.instrument; }
});
Object.defineProperty(exports, "isClientEnvironment", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isClientEnvironment; }
});
Object.defineProperty(exports, "isCompositeFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isCompositeFiber; }
});
Object.defineProperty(exports, "isHostFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isHostFiber; }
});
Object.defineProperty(exports, "isInstrumentationActive", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isInstrumentationActive; }
});
Object.defineProperty(exports, "isReactRefresh", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isReactRefresh; }
});
Object.defineProperty(exports, "isRealReactDevtools", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isRealReactDevtools; }
});
Object.defineProperty(exports, "isValidElement", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isValidElement; }
});
Object.defineProperty(exports, "isValidFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.isValidFiber; }
});
Object.defineProperty(exports, "mountFiberRecursively", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.mountFiberRecursively; }
});
Object.defineProperty(exports, "onCommitFiberRoot", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.onCommitFiberRoot; }
});
Object.defineProperty(exports, "patchRDTHook", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.patchRDTHook; }
});
Object.defineProperty(exports, "safelyInstallRDTHook", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.safelyInstallRDTHook; }
});
Object.defineProperty(exports, "secure", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.secure; }
});
Object.defineProperty(exports, "setFiberId", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.setFiberId; }
});
Object.defineProperty(exports, "shouldFilterFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.shouldFilterFiber; }
});
Object.defineProperty(exports, "traverseContexts", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.traverseContexts; }
});
Object.defineProperty(exports, "traverseFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.traverseFiber; }
});
Object.defineProperty(exports, "traverseProps", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.traverseProps; }
});
Object.defineProperty(exports, "traverseRenderedFibers", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.traverseRenderedFibers; }
});
Object.defineProperty(exports, "traverseState", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.traverseState; }
});
Object.defineProperty(exports, "unmountFiber", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.unmountFiber; }
});
Object.defineProperty(exports, "unmountFiberChildrenRecursively", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.unmountFiberChildrenRecursively; }
});
Object.defineProperty(exports, "updateFiberRecursively", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.updateFiberRecursively; }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function () { return chunkXUTTH4PR_cjs.version; }
});
